### PR TITLE
Combine pbr materials

### PR DESCRIPTION
**Pull request by Vonsen**
Combining AO, Metallic and roughness textures into one texture through compute shader works. It is able to run both mesh shader and ray tracing.

### CODE REVIEW

File: Assets/Shaders/CombineMaterial.comp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+layout(binding = 0, set = 0) uniform sampler2D 	u_RoughnessMaterial;
+layout(binding = 1, set = 0) uniform sampler2D 	u_MetallicMaterial;
+layout(binding = 2, set = 0) uniform sampler2D 	u_AOMaterial;
+
+layout(binding = 3, set = 0, rgba8) uniform writeonly image2D u_OutputImage;
+
+void main()
+{
+    // Working code
```
Comment (frozenbanana) [inline Assets/Shaders/CombineMaterial.comp]: Comment not adding any clarity consider removing it.

File: LambdaEngine/Source/Rendering/RenderAPI.cpp [hunk 2]
```diff
@@ -35,7 +35,7 @@ namespace LambdaEngine
 	{
 		GraphicsDeviceDesc deviceDesc = { };
 #ifdef LAMBDA_DEVELOPMENT
-		deviceDesc.Debug = true;
+		deviceDesc.Debug = false;
 #else
 		deviceDesc.Debug = false;
 #endif
```
Comment (frozenbanana) [inline LambdaEngine/Source/Rendering/RenderAPI.cpp:41]: Should it really be false in both cases?

Comment (frozenbanana) [inline LambdaEngine/Source/Rendering/RenderAPI.cpp]: Should it be false on both case?

File: LambdaEngine/Source/Resources/ResourceManager.cpp [hunk 3]
```diff
@@ -406,6 +419,213 @@ namespace LambdaEngine
...
 	}
 
+	void ResourceManager::CombineMaterials(Material* pMaterial, GUID_Lambda guid)
+	{
+		// Find largest texture size
+		uint32 largestHeight;
+		uint32 largestWidth;
+		largestWidth	= std::max(pMaterial->pMetallicMap->GetDesc().Width, std::max(pMaterial->pRoughnessMap->GetDesc().Width, pMaterial->pAmbientOcclusionMap->GetDesc().Width));
+		largestHeight	= std::max(pMaterial->pMetallicMap->GetDesc().Height, std::max(pMaterial->pRoughnessMap->GetDesc().Height, pMaterial->pAmbientOcclusionMap->GetDesc().Height));
+
```
Comment (frozenbanana) [inline LambdaEngine/Source/Resources/ResourceManager.cpp]: Consider assigning largestHeight and largestWidth directly instead. I think it is clear enough.

File: LambdaEngine/Source/Rendering/Core/Vulkan/CommandListVK.cpp [hunk 4]
```diff
@@ -470,6 +470,49 @@ namespace LambdaEngine
...
+		deferredBarrier.DestStages	= ConvertPipelineStageMask(dstStage);
+		if (!m_DeferredBarriers.IsEmpty())
+		{
+			for (DeferredImageBarrier& barrier : m_DeferredBarriers)
+			{
+				if (barrier.HasCompatableStages(deferredBarrier))
+				{
+					barrier.Barriers.EmplaceBack(imageBarrier);
+					break;
+				}
```
Comment (frozenbanana) [inline LambdaEngine/Source/Rendering/Core/Vulkan/CommandListVK.cpp:506]: Algorithm looks fine but consider using 

`if (std::find_if(m_DeferredBarriers.begin(), m_DeferredBarriers.end(), [deferredBarrier](const auto* pExcludedType) {
                    return barrier.HasCompatableStages(deferredBarrier);
                })) { ...`

**Review by frozenbanana — APPROVED** [review]
Looks good to me, though have not done any rendering work myself. I made a few comments to consider.

File: Assets/Shaders/CombineMaterial.comp (hunk 1 above)
Comment (Vonsen) [reply Assets/Shaders/CombineMaterial.comp]: fixed

File: LambdaEngine/Source/Resources/ResourceManager.cpp (hunk 3 above)
Comment (Vonsen) [reply LambdaEngine/Source/Resources/ResourceManager.cpp]: fixed

File: LambdaEngine/Source/Rendering/RenderAPI.cpp (hunk 2 above)
Comment (Vonsen) [reply LambdaEngine/Source/Rendering/RenderAPI.cpp]: Forgot to switch it back. It is necessary for both to be false when debugging using nsight graphics

**Pull request merged**